internal/app/ntpd/pkg/ntp/ntp.go

```diff
@@ -5,93 +5,110 @@
 package ntp
 
 import (
-	"errors"
+	"fmt"
 	"log"
 	"math/rand"
 	"syscall"
 	"time"
 
 	"github.com/beevik/ntp"
-)
-
-// https://access.redhat.com/solutions/39194
-// Using the above as reference for setting min/max
-const (
-	MaxPoll = 1000
-	MinPoll = 20
+	"github.com/hashicorp/go-multierror"
 )
 
 // NTP contains a server address
-// and the most recent response from a query
 type NTP struct {
-	Server   string
-	Response *ntp.Response
+	Server  string
+	MinPoll time.Duration
+	MaxPoll time.Duration
+	Retry   int
 }
 
 // NewNTPClient instantiates a new ntp client for the
 // specified server
-func NewNTPClient(server string) *NTP {
-	return &NTP{Server: server}
+func NewNTPClient(opts ...Option) (*NTP, error) {
+	ntp := defaultOptions()
+
+	var result *multierror.Error
+	for _, setter := range opts {
+		result = multierror.Append(setter(ntp))
+	}
+
+	return ntp, result.ErrorOrNil()
 }
 
 // Daemon runs the control loop for query and set time
 // We dont ever want the daemon to stop, so we only log
 // errors
 func (n *NTP) Daemon() (err error) {
-	rando := rand.New(rand.NewSource(time.Now().UnixNano()))
-	ticker := time.NewTicker(time.Duration(rando.Intn(MaxPoll)+MinPoll) * time.Second)
 
 	// Do an initial hard set of time to ensure clock skew isnt too far off
 	var resp *ntp.Response
-	resp, err = n.Query()
-	if err != nil {
+	if resp, err = n.Query(); err != nil {
 		log.Printf("error querying %s for time, %s", n.Server, err)
 		return err
 	}
-	n.Response = resp
 
-	if err = n.SetTime(resp); err != nil {
+	if err = setTime(resp.Time); err != nil {
 		return err
 	}
 
+	var randSleep time.Duration
 	for {
-		<-ticker.C
-		// Set some variance with how frequently we poll ntp servers
-		resp, err = n.Query()
-		if err != nil {
+		// Set some variance with how frequently we poll ntp servers.
+		// This is based on rand(MaxPoll) + MinPoll so we wait at least
+		// MinPoll.
+		randSleep = time.Duration(rand.Intn(int(n.MaxPoll.Seconds()))) * time.Second
```

Comment:
I'm a big fan of exponential distribution to manage such intervals (https://godoc.org/math/rand#ExpFloat64). 

This could be different from MinPoll/MaxPoll setup, but should provide more natural distribution of poll intervals across cluster. 

Not really blocking this PR, I think what we have here with uniform distribution looks good.

Reply:
Its probably worth us building out a library for generic backoff strategies. We use them in a number of places. Obviously not for this PR, but while we are on the topic of polling I figured I'd throw in my 2 cents.